### DISCUSSION TITLE
question regarding simulation_frequency

### DISCUSSION THREAD

**Discussion by JianmingTONG** [eleurent/highway-env#229] [Q&A] [answered]
Hi, I have found that in the file ```/highway_env/envs/common/abstract.py```, there are two parameters related to the frequency, including **simulation_frequency** and **policy_frequency**. Based on my experiments, the lower ratio of **simulation_frequency** to **policy_frequency**, the faster the execution of the environment step function, because the iteration step of ```road.act()``` is reduced. 

However, I'm not so clear about the meaning of the two parameters. Might I request some insight from you on these parameters?

And are there value ranges of both parameters to make the simulation run accurately? For example, does it impact the accuracy if I set both two parameters as 1.

Best regards
Jianming

**Comment by eleurent** [comment, marked as answer]
Hi, 
* `simulation_frequency` is the frequency used for the Euler integration of the dynamics, i.e. the physical simulation. A high simulation frequency will give more accurate results, at the price of an increased computational load (since more intermediate timesteps need to be computed). I haven't experimented recently, but I think that the simulation tends to become unstable under 5Hz. There is no upper bound, but at some point the benefits in terms of accuracy become negligible. I'd say [5Hz, 15Hz] is a good range.
* `policy_frequency`, on the other hand, is the frequency at which the agent can take decisions. It cannot be higher than the simulation frequency, and this upper bound corresponds to the standard case where the agent provides a control input for every simulated frame, which makes sense for e.g. low-level control (steering/throttle). But for high-level decisions, such as lane changes, it can be more sensible to have temporally extended actions: if a lane change takes at least 1 second to execute, we do not need to take decisions at a higher frequency. Increasing the policy frequency means that a call to the `step()` method will be faster, but in turn the corresponding simulated duration will be shorter.

By default, for low-level control tasks, I use the values of `simulation_frequency=policy_frequency=15`, and for high-level decision tasks I use `simulation_frequency=15` and `policy_frequency=1`

**Comment by freemedom** [comment]
https://github.com/Farama-Foundation/HighwayEnv/blob/0520a88de118040be0d79b7b25d16fc5c1bde8bb/highway_env/envs/common/abstract.py#L249
the lower ratio of simulation_frequency to policy_frequency, the smaller value of `frames`(simulated frame)

    def _simulate(self, action: Optional[Action] = None) -> None:
        """Perform several steps of simulation with constant action."""
        frames = int(self.config["simulation_frequency"] // self.config["policy_frequency"])
        for frame in range(frames):
            # Forward action to the vehicle

If policy_frequency is higher than the simulation_frequency, the `frames` will be 0, it will not enter the for loop, ego-vehicle and other vehicles will not move.

---
https://github.com/Farama-Foundation/HighwayEnv/blob/0520a88de118040be0d79b7b25d16fc5c1bde8bb/highway_env/envs/common/abstract.py#L254
The greater the probability of divisible(//)

---

            if action is not None \
                    and not self.config["manual_control"] \
                    and self.steps % int(self.config["simulation_frequency"] // self.config["policy_frequency"]) == 0:
                self.action_type.act(action)#Once called, the agent/ego-vehicle takes decision once

            self.road.act()

The greater the probability of ego-vehicle "act/take decisions", more timely change lanes acceleration deceleration.(self.road act every loop, but self.action_type/ego-vehicle does not necessarily)

**Comment by freemedom** [comment]
https://github.com/Farama-Foundation/HighwayEnv/blob/0520a88de118040be0d79b7b25d16fc5c1bde8bb/highway_env/envs/common/abstract.py#L258
The greater the simulation_frequency, the less dt(timestep [s]) the self.road step.

simulation_frequency affects "the environment's running speed/the vehicle's visual driving speed in the pygame window", policy_frequency has no effect. The larger the simulation_frequency, the less dt of each step, the slower the running speed, and the more real seconds corresponding to one simulated second.

**Comment by freemedom** [comment]
> it can be more sensible to have temporally extended actions

I don't understand what this sentence means?

**Reply by eleurent** [reply]
In some cases you may want actions to ask for a long time (e.g. perform a full maneuver over several seconds), typically for high-level decision tasks (but not for low-level control, where a high frequency is required). The benefit of having longer actions is that it makes exploration easier, as a few actions can take you to a distant state. The drawback is that it limits the set of achievable trajectories, and can make the best achievable policy suboptimal.
